codeU3.py

```diff
@@ -0,0 +1,222 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Sat Jun 10 09:17:26 2017
+Exercise 3, version 1
+@author: Luiza
+"""
+
+from collections import deque
+import unittest
+import numpy as np
+
+def getPrefixes(word):
+
+    '''Returns the list of all prefixes of given word
+    including the word itself
+    '''
+
+    return set([word[0:n] for n in range(1, len(word)+1)])
+
+class Dictionary(object):
+
+    '''Stores lower-case vocabulary and all its prefixes as a set
+    Two functions isWord and isPrefix are defined to check if given word/prefix is found
+    Constructor needs only a non-empty word-list, prefixes found automatically
+    '''
+
+    def __init__(self, word_list):
+        if not word_list:
+            assert ValueError, 'Please provide non-empty list of words'
+        word_list = [w.lower() for w in word_list if w]
+        self.max_word_len = reduce(max, map(len, word_list))
+        self.words = set(word_list)
+        self.word_prefixes = set()
+        for word in self.words:
+            self.word_prefixes.update(getPrefixes(word))
+
+    def isWord(self, word):
+
+        '''Returns whether the word presents in the vocabulary
+        the function is case-invariant
+        '''
+
+        if not isinstance(word, str):
+            assert TypeError, 'Input should be a string'
+            return False
+        if word.lower() in self.words:
+            return True
+        return False
+
+    def isPrefix(self, prefix):
+
+        '''Returns whether the string is the prefix of one of the words in vocabulary
+        the function is case-invariant
+        '''
+
+        if not isinstance(prefix, str):
+            assert TypeError, 'Input should be a string'
+            return
+        if prefix.lower() in self.word_prefixes:
+            return True
+        return False
+
+class TestDictionary(unittest.TestCase):
+
+    '''Testing dictionary basic functions
+    '''
+
+    def setUp(self):
+        self.d = Dictionary(['cats', 'dogs', 'cars', ' ', ''])
+
+    def test_empty_input(self):
+        self.assertFalse(self.d.isWord('  '))
+        self.assertFalse(self.d.isPrefix(''))
+
+    def test_case(self):
+        self.assertTrue(self.d.isWord('Cats'))
+        self.assertTrue(self.d.isPrefix('DOGS'))
+
+    def test_normal_input(self):
+        self.assertTrue(self.d.isWord('cats'))
+        self.assertTrue(self.d.isPrefix('car'))
+
+        self.assertFalse(self.d.isWord('cat'))
+        self.assertFalse(self.d.isPrefix('cs'))
+
+    def test_nonstring_input(self):
+        self.assertRaises(self.d.isWord(['cat', 'cats']))
+        self.assertRaises(self.d.isPrefix(['cat', 'cats']))
+
+class CharGrid(object):
+
+    '''Given (n by m) grid of letters and a dictionary

```

Comment:
Thanks for writing a class-level docstring as well as the function docstrings - helps a lot with reading and understanding any code base!